cirq/contrib/qasm_import/_parser.py

```diff
@@ -35,7 +36,42 @@ def __init__(self, supported_format: bool, qelib1_include: bool,
         self.circuit = c
 
 
-class QasmParser(object):
+class QasmGate:
+
+    def __init__(self, qasm_gate: str, cirq_gate: cirq.Gate, num_args: int):
+        self.qasm_gate = qasm_gate
+        self.cirq_gate = cirq_gate
+        self.num_args = num_args
+
+    def validate_args(self, args, lineno):
+        if len(args) != self.num_args:
+            raise QasmException(
+                "{} only takes {} arg(s) (qubits and/or registers), "
+                "got: {}, at line {}".format(self.qasm_gate, self.num_args,
+                                             len(args), lineno))
+
+    def on(self, args: List[List[cirq.Qid]],
+           lineno: int = 0) -> Iterable[cirq.Operation]:
+        self.validate_args(args, lineno)
+        reg_size = 1
+        for reg in args:
+            if len(reg) > 1 and len(reg) != reg_size:
+                if reg_size == 1:
+                    reg_size = len(reg)
+                else:
+                    raise QasmException(
+                        "Non matching quantum registers of length {} "
+                        "at line {}".format([len(reg) for reg in args], lineno))
+
+        for i in range(reg_size):
+            # because of mixed quantum register/qubit support there can
+            # be single qubit registers and longer qubit registers
+            # for single qubit registers qreg[0], for longer ones qreg[i]
+            qubits = [qreg[min(len(qreg) - 1, i)] for qreg in args]
```

Comment:
Is this a correct interpretation of the input? It looks like it will replace malformed expressions like `q[999999999999999999999999999]` with `q[max_admissible_index]` instead of producing an error.

Reply:
Because we check for it above, the registers will be either the same size or single qubits. I will replace this `min` logic with a more understandable loop. 

That validation for out of bounds is a real bug though, thank you for catching it. It should happen here at line 210 in `p_arg_bit` - currently a `q[999999999999999999999999999]` would create a `NamedQubit` with that name in a register without any issues. Will fix it.